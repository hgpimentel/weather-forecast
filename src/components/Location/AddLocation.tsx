import React, { useState, useContext } from "react";
import { ColumnContainer } from "../../components";
import LocationContext, { Location } from "../../store/location-context";
import styled from "styled-components";

const Form = styled.form`
  width: 100%;
  display: flex;
  flex-direction: column;

  span {
    width: 100px;
  }

  & > * {
    margin-bottom: 0.5rem;
    display: flex;
  }

  button {
    width: 100px;
    text-align: center;
    margin-top: 0.5rem;
    background-color: teal;
    color: white;
    display: inline-block;
    padding: 5px;
    border-radius: 5px;
    border: none;
    cursor: pointer;

    &:hover {
      background-color: rgba(0, 128, 128, 0.75);
    }
  }
`;

const AddLocation: React.FC<{ closeModal: () => void }> = ({ closeModal }) => {
  const [city, setCity] = useState<string>("");
  const [country, setCountry] = useState<string>("");
  const [lat, setLat] = useState<number>(0.0);
  const [long, setLong] = useState<number>(0.0);
  const { onLocationChange } = useContext(LocationContext);

  const onSubmit = (e: React.SyntheticEvent) => {
    e.preventDefault();
    const lsItem = localStorage.getItem("locations");
    const locations = lsItem ? (JSON.parse(lsItem) as Location[]) : null;
    const location: Location = {
      city,
      country,
      lat,
      long,
    };

    if (locations) {
      locations.push(location);
      localStorage.setItem("locations", JSON.stringify(locations));
    } else {
      localStorage.setItem("locations", JSON.stringify([location]));
    }

    onLocationChange(location);
    closeModal();
  };

  const canSubmit = city && country && lat && long;

  return (
    <ColumnContainer>
      <div>
        <h3>New location</h3>
      </div>
      <Form>
        <div>
          <span>City</span>
          <input
            type="text"
            id="city"
            name="city"
            value={city}
            onChange={(e) => setCity(e.target.value)}
          />
        </div>
        <div>
          <span>Country</span>
          <input
            type="text"
            id="country"
            name="country"
            value={country}
            onChange={(e) => setCountry(e.target.value)}
          />
        </div>
        <div>
          <span>Latitude</span>
          <input
            type="number"
            id="lat"
            name="lat"
            value={lat}
            onChange={(e) => setLat(e.target.valueAsNumber)}
          />
        </div>
        <div>
          <span>Longitude</span>
          <input
            type="number"
            id="long"
            name="long"
            value={long}
            onChange={(e) => setLong(e.target.valueAsNumber)}
          />
        </div>
        <button type="submit" disabled={!canSubmit} onClick={onSubmit}>
          Add
        </button>
      </Form>
    </ColumnContainer>
  );
};

export default AddLocation;
