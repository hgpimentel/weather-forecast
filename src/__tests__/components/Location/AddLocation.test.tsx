import renderer from "react-test-renderer";
import AddLocation from "../../../components/Location/AddLocation";
import { render, screen, fireEvent } from "@testing-library/react";

describe("Add location component", () => {
  it("renders correctly", () => {
    const component = renderer
      .create(<AddLocation closeModal={jest.fn()} />)
      .toJSON();

    expect(component).toMatchSnapshot();
  });

  it("has button disabled", () => {
    render(<AddLocation closeModal={jest.fn()} />);

    expect(screen.getByRole("heading")).toHaveTextContent("New location");

    const button = screen.getByRole("button");

    expect(button).toHaveTextContent("Add");
    expect(button).toBeDisabled();
  });

  it("has button enabled", () => {
    render(<AddLocation closeModal={jest.fn()} />);

    screen.queryAllByDisplayValue("").forEach((elem) => {
      fireEvent.change(elem, {
        target: { value: "Text" },
      });
    });

    screen.queryAllByDisplayValue(0).forEach((elem) => {
      fireEvent.change(elem, {
        target: { value: "10.010101" },
      });
    });

    const button = screen.getByRole("button");
    expect(button).not.toBeDisabled();
  });
});
