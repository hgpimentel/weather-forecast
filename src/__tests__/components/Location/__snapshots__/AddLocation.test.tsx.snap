// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add location component renders correctly 1`] = `
<div
  className="sc-bdnxRM eoiijl"
>
  <div>
    <h3>
      New location
    </h3>
  </div>
  <form
    className="sc-jrsJWt hSJTmk"
  >
    <div>
      <span>
        City
      </span>
      <input
        id="city"
        name="city"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div>
      <span>
        Country
      </span>
      <input
        id="country"
        name="country"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div>
      <span>
        Latitude
      </span>
      <input
        id="lat"
        name="lat"
        onChange={[Function]}
        type="number"
        value={0}
      />
    </div>
    <div>
      <span>
        Longitude
      </span>
      <input
        id="long"
        name="long"
        onChange={[Function]}
        type="number"
        value={0}
      />
    </div>
    <button
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      Add
    </button>
  </form>
</div>
`;
